src/dsp_tools/utils/xmlupload/xmlupload.py

```diff
@@ -81,53 +72,87 @@ def xmlupload(
     default_ontology, root, shortcode = validate_and_parse_xml_file(
         input_file=input_file,
         imgdir=imgdir,
-        preprocessing_done=preprocessing_done,
+        preprocessing_done=config.preprocessing_done,
     )
 
-    # determine save location that will be used for diagnostic info if the xmlupload is interrupted
-    save_location, server_as_foldername, timestamp_str = determine_save_location_of_diagnostic_info(
+    config = config.with_specific_save_location(
         server=server,
-        proj_shortcode=shortcode,
+        shortcode=shortcode,
         onto_name=default_ontology,
     )
-    logger.info(f"save_location='{save_location}'")
 
     # start metrics
     metrics: list[MetricRecord] = []
-    preparation_start = datetime.now()
 
     # establish connection to DSP server
-    con = login(server=server, user=user, password=password, dump=dump)
+    con = login(server=server, user=user, password=password, dump=config.dump)
     sipi_server = Sipi(sipi, con.get_token())
 
-    proj_context = _get_project_context_from_server(connection=con)
-
-    permissions = _extract_permissions_from_xml(root, proj_context)
-    resources = _extract_resources_from_xml(root, default_ontology)
-
-    permissions_lookup, resclass_name_2_type = _get_project_permissions_and_classes_from_server(
-        server_connection=con,
-        permissions=permissions,
-        shortcode=shortcode,
+    resources, permissions_lookup, resclass_name_2_type, stash = _prepare_upload(
+        root, con, default_ontology, shortcode, config.verbose, metrics
```

Comment:
I always prefer keyword arguments, because it makes the program more robust against involuntary changes, e.g. if you move the order of the parameters in the function signature